src/util/ringdelaybuffer.cpp

```diff
@@ -0,0 +1,211 @@
+/*
+* $Id$
+* Portable Audio I/O Library
+* Ring Buffer utility.
+*
+* Author: Phil Burk, http://www.softsynth.com
+* modified for SMP safety on OS X by Bjorn Roche.
+* also allowed for const where possible.
+* modified for multiple-byte-sized data elements by Sven Fischer
+*
+* This program is distributed with the PortAudio Portable Audio Library.
+* For more information see: http://www.portaudio.com
+* Copyright (c) 1999-2000 Ross Bencina and Phil Burk
+*
+* Permission is hereby granted, free of charge, to any person obtaining
+* a copy of this software and associated documentation files
+* (the "Software"), to deal in the Software without restriction,
+* including without limitation the rights to use, copy, modify, merge,
+* publish, distribute, sublicense, and/or sell copies of the Software,
+* and to permit persons to whom the Software is furnished to do so,
+* subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be
+* included in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
+* ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+* CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+* WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+/*
+* The text above constitutes the entire PortAudio license; however,
+* the PortAudio community also makes the following non-binding requests:
+*
+* Any person wishing to distribute modifications to the Software is
+* requested to send the modifications to the original developer so that
+* they can be incorporated into the canonical version. It is also
+* requested that these non-binding requests be included along with the
+* license above.
+*/
+
+#include "ringdelaybuffer.h"
+
+#include "util/sample.h"
+
+RingDelayBuffer::RingDelayBuffer(SINT bufferSize)
+        : m_fullFlag(false),
+          m_firstInputBuffer(true),
+          m_readPos(0),
+          m_writePos(0),
+          m_buffer(bufferSize) {
+    // Set the ring delay buffer items to 0.
+    m_buffer.fill(0);
+}
+
+SINT RingDelayBuffer::read(CSAMPLE* pBuffer, const SINT numItems) {
+    const SINT available = getReadAvailable();
+    SINT itemsToRead = numItems;
+
+    VERIFY_OR_DEBUG_ASSERT(itemsToRead <= available) {
+        itemsToRead = available;
```

Comment:
I think you should zero out the remaining buffer. It does not really make sense to put that work to the using code. 
We may also consider to just return without touching the buffer, or zero out the whole buffer.

Reply:
Could be. To me, the version that it will be just returned from the function without touching the buffer looks more intuitively. The zero value will be returned and the user code can handle this situation on its own. IMO clearing the buffer looks like a little bit of unexpected behaviour to me, but that is just my view about it.

Reply:
Ok